tree-walk-xl.go

```diff
@@ -138,23 +140,37 @@ func (xl xlObjects) treeWalk(bucket, prefixDir, entryPrefixMatch, marker string,
 				// recursing into "four/"
 				markerArg = markerBase
 			}
-			*count--
 			prefixMatch := "" // Valid only for first level treeWalk and empty for subdirectories.
-			if !xl.treeWalk(bucket, pathJoin(prefixDir, entry), prefixMatch, markerArg, recursive, send, count, isLeaf) {
-				return false
+			if i == len(entries)-1 && stackDepth == 0 {
+				isEnd = true
 			}
+			stackDepth++
+			xl.treeWalk(bucket, pathJoin(prefixDir, entry), prefixMatch, markerArg, recursive, isLeaf, treeWalkCh, doneCh, stackDepth, isEnd)
+			stackDepth--
 			continue
 		}
-		*count--
-		if !send(treeWalkResult{entry: pathJoin(prefixDir, entry)}) {
-			return false
+		var isEOF bool
+		if stackDepth == 0 && i == len(entries)-1 {
+			isEOF = true
+		} else if i == len(entries)-1 && isEnd {
+			isEOF = true
 		}
+		select {
+		case <-doneCh:
+			if stackDepth == 0 {
+				close(treeWalkCh)
+				return
+			}
+		case treeWalkCh <- treeWalker{entry: pathJoin(prefixDir, entry), end: isEOF}:
```

Comment:
for this structure:
```
root/
├── a
│   └── file.1
└── file.2
```

if the go routine receives on doneCh (because of timeout) when sending file.1, the treeWalk function returns and the caller treeWalk tries to send file.2 into the channel and blocks.